statsmodels/base/data.py

```diff
@@ -91,11 +91,25 @@ def __getstate__(self):
     def __setstate__(self, d):
         if "restore_design_info" in d:
             # NOTE: there may be a more performant way to do this
-            from patsy import dmatrices
-            depth = 1  # hmm, have to have same eval env in calling ns
-            data = d['orig_endog'].join(d['orig_exog'])
-            _, design = dmatrices(d['formula'], data, eval_env=depth,
-                                  return_type='dataframe')
+            from patsy import dmatrices, PatsyError
+            exc = []
+            try:
+                data = d['frame']
+            except KeyError:
+                data = d['orig_endog'].join(d['orig_exog'])
+
+            for depth in [2, 3, 1, 0, 4]:
+                try:
+                    _, design = dmatrices(d['formula'], data, eval_env=depth,
+                                          return_type='dataframe')
+                    break
+                except PatsyError as e: #NameError as e:
+                    #print('not in depth %d' % depth)
```

Comment:
should we add a `warning` about which eval depth is used?
This could create problems if there are several function with the same name in different namespaces, I guess.